File: ios/SessionAttributes/SessionAttributesEngine.swift
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import Foundation

@objc public class SessionAttributesEngine: NSObject {
    @objc public static let shared = SessionAttributesEngine()

    private let store = SessionAttributesStore.shared
    private let collector = SessionAttributesCollector.shared

    private override init() {}

    @objc public func setEnabled(_ serverUrl: String, enabled: Bool) {
        store.setEnabled(enabled, for: serverUrl)
    }

    @objc public func removeServer(_ serverUrl: String) {
        store.removeState(for: serverUrl)
    }

    @objc public func setManifest(_ serverUrl: String, manifestJson: String) {
        let fields = decodeArray(manifestJson).compactMap { parseField($0) }
        guard !fields.isEmpty else {
            removeServer(serverUrl)
            return
        }
        store.setManifest(fields, for: serverUrl)
    }

    @objc public func upsertManifestField(_ serverUrl: String, fieldJson: String) {
        guard let dictionary = decodeObject(fieldJson), let parsed = parseField(dictionary) else {
            return
        }
        store.upsertField(parsed, for: serverUrl)
    }

    @objc public func removeManifestField(_ serverUrl: String, name: String) {
        store.removeField(name, for: serverUrl)
    }

    @objc public func setStableValues(_ valuesJson: String) {
        guard let data = valuesJson.data(using: .utf8),
              let decoded = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return
        }
        let values = decoded.compactMapValues { value -> String? in
            if let string = value as? String {
                return string
            }
            return nil
        }
        store.setStableValues(values)
    }

    @objc public func getOutboundHeader(_ serverUrl: String) -> String? {
        guard let state = store.loadState(for: serverUrl), state.enabled, !state.manifest.isEmpty else {
            return nil
        }

        let now = Date().timeIntervalSince1970 * 1000
        let lastSentAt = state.lastSentAt
        var payload: [String: String] = [:]
        var updatedLastSentAt = lastSentAt

        for field in state.manifest {
            let lastSent = lastSentAt[field.name]
            let shouldSend = lastSent == nil || field.ttl_seconds == 0 || (now - lastSent!) >= Double(field.ttl_seconds) * 1000
            if !shouldSend {
                continue
            }

            let value = collector.collect(field.name, serverUrl: serverUrl)
            guard !value.isEmpty else {
                continue
            }

            payload[field.name] = value
            updatedLastSentAt[field.name] = now
        }

        guard !payload.isEmpty else {
            return nil
        }

        store.updateLastSentAt(updatedLastSentAt, for: serverUrl)

        guard let jsonData = try? JSONSerialization.data(withJSONObject: payload, options: []),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            return nil
        }

        return Data(jsonString.utf8).base64EncodedString()
    }

    private func decodeArray(_ json: String) -> [[String: Any]] {
        guard let data = json.data(using: .utf8),
              let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }
        return array
    }

    private func decodeObject(_ json: String) -> [String: Any]? {
        guard let data = json.data(using: .utf8) else {
            return nil
        }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private func parseField(_ dictionary: [String: Any]) -> SAField? {
        guard let name = dictionary["name"] as? String,
              let type = dictionary["type"] as? String else {
            return nil
        }
        let ttl = dictionary["ttl_seconds"] as? Int ?? (dictionary["ttl_seconds"] as? Double).map { Int($0) } ?? 0
        let grace = dictionary["grace_period_seconds"] as? Int ?? (dictionary["grace_period_seconds"] as? Double).map { Int($0) } ?? 0
        return SAField(name: name, type: type, ttl_seconds: ttl, grace_period_seconds: grace)
    }
}

File: android/src/main/java/com/mattermost/networkclient/sessionattributes/SessionAttributesEngine.kt
package com.mattermost.networkclient.sessionattributes

import android.content.Context
import android.util.Base64
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject

object SessionAttributesEngine {
    private lateinit var context: Context

    private val store: SessionAttributesStore by lazy { SessionAttributesStore(context) }
    private val collector: SessionAttributesCollector by lazy { SessionAttributesCollector(context, store) }

    fun init(appContext: Context) {
        if (!::context.isInitialized) {
            context = appContext.applicationContext
        }
    }

    fun setEnabled(serverUrl: String, enabled: Boolean) {
        store.setEnabled(serverUrl, enabled)
    }

    fun removeServer(serverUrl: String) {
        store.removeState(serverUrl)
    }

    fun setManifest(serverUrl: String, manifestJson: String) {
        val manifest = try {
            JSONArray(manifestJson)
        } catch (e: Exception) {
            Log.w("NetworkClient", "Discarding malformed manifest for $serverUrl: ${e.message}")
            null
        }
        val fields = mutableListOf<SAField>()
        if (manifest != null) {
            for (i in 0 until manifest.length()) {
                SAField.fromJson(manifest.getJSONObject(i))?.let { fields.add(it) }
            }
        }
        if (fields.isEmpty()) {
            removeServer(serverUrl)
            return
        }
        store.setManifest(serverUrl, fields)
    }

    fun upsertManifestField(serverUrl: String, fieldJson: String) {
        val field = try {
            JSONObject(fieldJson)
        } catch (e: Exception) {
            Log.w("NetworkClient", "Ignoring malformed manifest field for $serverUrl: ${e.message}")
            return
        }
        SAField.fromJson(field)?.let { store.upsertField(serverUrl, it) }
    }

    fun removeManifestField(serverUrl: String, name: String) {
        store.removeField(serverUrl, name)
    }

    fun setStableValues(valuesJson: String) {
        val json = try {
            JSONObject(valuesJson)
        } catch (e: Exception) {
            Log.w("NetworkClient", "Ignoring malformed stable values: ${e.message}")
            return
        }
        val values = mutableMapOf<String, String>()
        json.keys().forEach { key ->
            values[key] = json.optString(key, "")
        }
        store.setStableValues(values)
    }

    fun getOutboundHeader(serverUrl: String): String? {
        if (!::context.isInitialized) {
            Log.w("NetworkClient", "Session attributes requested before init")
            return null
        }

        val state = store.loadState(serverUrl) ?: return null
        if (!state.enabled || state.manifest.isEmpty()) {
            return null
        }

        val now = System.currentTimeMillis()
        val payload = JSONObject()

        for (field in state.manifest) {
            val lastSent = state.lastSentAt[field.name]
            val shouldSend = lastSent == null || field.ttlSeconds == 0 ||
                (now - lastSent) >= field.ttlSeconds * 1000L
            if (!shouldSend) {
                continue
            }

            val value = collector.collect(field.name, serverUrl)
            if (value.isEmpty()) {
                continue
            }

            payload.put(field.name, value)
            state.lastSentAt[field.name] = now
        }

        if (payload.length() == 0) {
            return null
        }

        store.saveState(serverUrl, state)

        return Base64.encodeToString(payload.toString().toByteArray(), Base64.NO_WRAP)
    }
}
